### ISSUE TITLE
Waiting executions doesn't follow FIFO

### ISSUE THREAD

**Issue by JamieChe** [spinnaker/spinnaker#6373]
### Issue Summary:
Waiting executions doesn't follow FIFO

### Environment:
Spinnaker 1.21.4, Spinnaker 1.25

### Feature Area:
Orca queue

### Description:
The sequence of queued executions, which should follow FIFO,  gets messed up when two conditions are met
1. The pipeline has branches.  When one branch completes, the running execution is waiting for other branches to complete.
2. Pipeline disables concurrent executions. So several pending executions are waiting in a queue when pipeline has running executions. 

Checked redis queue, the order of 01F0WQV9NR65J9F9N2FWN8FRVK is changed from oldest to newest. And the attempts attribute get +1
```
127.0.0.1:6379> LRANGE orca.pipeline.queue.8b4f6fbf-df5b-4ac9-9b0e-a63b3dd38c97 0 10
1) "{\"kind\":\"startExecution\",\"executionType\":\"PIPELINE\",\"executionId\":\"01F0WQVEY9XK52Y47W2DD1F91W\",\"application\":\"issuedebug\",\"attributes\":[{\"kind\":\"attempts\",\"attempts\":1}]}"
2) "{\"kind\":\"startExecution\",\"executionType\":\"PIPELINE\",\"executionId\":\"01F0WQVCB8EPRJAJNQ55VF3QTN\",\"application\":\"issuedebug\",\"attributes\":[{\"kind\":\"attempts\",\"attempts\":1}]}"
3) "{\"kind\":\"startExecution\",\"executionType\":\"PIPELINE\",\"executionId\":\"01F0WQV9NR65J9F9N2FWN8FRVK\",\"application\":\"issuedebug\",\"attributes\":[{\"kind\":\"attempts\",\"attempts\":1}]}"
127.0.0.1:6379> 
127.0.0.1:6379> 
127.0.0.1:6379> LRANGE orca.pipeline.queue.8b4f6fbf-df5b-4ac9-9b0e-a63b3dd38c97 0 10
1) "{\"kind\":\"startExecution\",\"executionType\":\"PIPELINE\",\"executionId\":\"01F0WQV9NR65J9F9N2FWN8FRVK\",\"application\":\"issuedebug\",\"attributes\":[{\"kind\":\"attempts\",\"attempts\":2}]}"
2) "{\"kind\":\"startExecution\",\"executionType\":\"PIPELINE\",\"executionId\":\"01F0WQVEY9XK52Y47W2DD1F91W\",\"application\":\"issuedebug\",\"attributes\":[{\"kind\":\"attempts\",\"attempts\":1}]}"
3) "{\"kind\":\"startExecution\",\"executionType\":\"PIPELINE\",\"executionId\":\"01F0WQVCB8EPRJAJNQ55VF3QTN\",\"application\":\"issuedebug\",\"attributes\":[{\"kind\":\"attempts\",\"attempts\":1}]}"
```
Checked redis message, looks like the "completeExecution" message of current running execution may be related to re-order.  But we have no idea why it pops out the oldest waiting execution and pushes it back. 
```
{"kind":"completeExecution","executionType":"PIPELINE","executionId":"01F0WQV6WFRXCS2JQ5C8X043WP","application":"issuedebug","attributes":[{"kind":"attempts","attempts":3}]}
{"kind":"runTask","executionType":"PIPELINE","executionId":"01F0WQV6WFRXCS2JQ5C8X043WP","application":"issuedebug","stageId":"01F0WQV6WTS33MK59K55FBEQ3S","taskId":"1","taskType":"com.netflix.spinnaker.orca.pipeline.tasks.WaitTask","attributes":[{"kind":"attempts","attempts":1}],"ackTimeoutMs":600000}
```

We met this issue since 2019 (previous issue https://github.com/spinnaker/spinnaker/issues/4587. ) Now it still exists in 1.21, even latest 1.25. We also tried mysql as queue,  not working. 

### Steps to Reproduce:
1. create one pipeline, with two wait stages as two branches. One wait stage1 set like 10 secs, the other wait stage2 set 300 secs
2. disable concurrent execution of this pipeline
3. run this pipeline 5 times, to make sure one running and 4 waiting. The 4 waiting executions should follow FIFO in redis queue at the very beginning (with attempt=1)
4. When wait stage1 complete but waitStage2 is still running, the "kind":"completeExecution" will be created in orca message. 
5. Now monitor the waiting 4 executions from both UI and redis queue, you can see the re-order. 

### Additional Details:

---

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by JamieChe** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by deverton** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by JamieChe** [comment]
@spinnakerbot remove-label stale

**Comment by JamieChe** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by JamieChe** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by stodmis** [comment]
Hello Spinnaker community,
The same problem is still present in 1.26.7 version. All executions in pending status within a pipeline don't follow FIFO rule (i.e. if the certain pipeline has a queue from executions with NOT_STARTED status so Spinnaker doesn't launch the oldest execution at first from the line).  Can someone tell me if this issue will be solved in the future versions?

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by arjunmenon** [comment]
@JamieChe Hey, did you ever get around in resolving this?

**Comment by arjunmenon** [comment]
@spinnakerbot remove-label to-be-closed
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by jimmycasey** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
"to-be-closed" has not been applied, and cannot be removed.

**Comment by rjalander** [comment]
I have created a PR to fix this issue, can someone help to review this and approve.
https://github.com/spinnaker/orca/pull/4356

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by rjalander** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by rjalander** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'to-be-closed' and hasn't been updated  in 45 days, so we are closing it. You can always reopen this issue if needed.

**Comment by rjalander** [comment]
Fixed with the PR https://github.com/spinnaker/orca/pull/4356

**Comment by qunayu** [comment]
Hey folks, we saw this happen again in another rare race condition situation. 

This can still happen (albeit much rarer and inconsistent) where your pipeline ends up multiple branches, which means each one of those can push a `CompleteExecution` message. The queue de-dupes on the body of the message (used to generate the fingerprint), so if multiple of these are pushed, they’ll be deduped one by one and only one `CompleteExecution` message will remain in the messages table.

This works well usually, unless - the queue polling happens to pick up the unique `CompleteExecution` message, then in the 1-20 ms that it’s being worked on, another `CompleteExecution` message is pushed. Because the other message was picked up, this message is inserted as if none currently exist. The first message gets worked on successfully, completing the current execution and lets the following one run by sending a `StartWaitingExecutions` message. The second message gets picked up in the next polling cycle, then sends another `StartWaitingExecutions` , which then pops the oldest waiting execution. It's then pushed back onto the queue because it can’t currently run. This disrupts the original order, and means the oldest execution (next in line to run) becomes the last one in line to run.

This happened to us where the first `CompleteExecution` message was picked up in the same millisecond that the second `CompleteExecution` was pushed into the DB.

We also saw the line `log.info("Execution ${execution.id} already completed with ${execution.status} status")` being logged, which means at the time the second message was worked on, it was already aware that the execution has been completed by the handler in another thread/process.

We're thinking of simply terminating the handler at this point https://github.com/spinnaker/orca/blob/9c8126b7e9234acb9ad2e1b19f1c4efae5f66d36/orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteExecutionHandler.kt#L63, when it sees the execution has already been marked as complete. 

Any thoughts? Do we think this would be a safe change?
